### PR TITLE
[Frontend] Add support for `jax.jit` and `jax.grad` on top of `qjit` functions

### LINKED ISSUE #15: Add `jax.jit` and JAX transformation integration
**Context**

There are two ways we could support 'external' JAX integration - that is, the ability to:

- include `@qjit` _inside_ `@jax.jit` hybrid functions, and
- call JAX transformations such as `jax.grad` and `jax.vmap` directly on `qjit` functions.

To do this, we could either use:

1. [Custom derivative rules](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html) to register the `@qjit` decorator with JAX, alongside [`pure_callback`](https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html) to allow for an XLA compiled program to 'callback' to Python to execute the `qjit` compiled program. 

2. [Custom JAX primitives](https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html#defining-new-jax-primitives) that can _directly_ make use of XLA [`CustomCalls`](https://www.tensorflow.org/xla/custom_call) to call arbitrary C++ code (see https://dfm.io/posts/extending-jax/)

Longer term, (2) is likely a better solution. Option (1), as it requires a Python callback, may introduce unneeded overhead.

However, Option (1) is quite easy to do, and might make sense to do first, since it enables the same user-facing behaviour.

<details>
  <summary><strong>Option (1) prototype/proof-of-concept (click to show)</strong></summary>
Below is a proof-of-concept of Option (1):

```python
import catalyst
import jax
import jax.numpy as jnp
def _make_jax_jit(func, shapes):
    """Use jax.pure_callback to make a JAX-JIT compatible QJIT function.

    Args:
        func (QJIT): the QJIT function to make jax-jit compatible
        shapes (Sequence[ShapeDtypeStruct]): list of shapes and dtypes to
            expect from the output of the QJIT function
    """
    qjit_func = catalyst.qjit(lambda args: func(*args))

    def wrapper(*args):
        # NOTE: at the moment, it appears that QJIT is not vectorized.
        # Can we support vectorization?

        if args[0].ndim > 1:
            # batch dimension is present. Hardcoding in only *one* batch dim.
            res = [jax.pure_callback(qjit_func, shapes, [a], vectorized=False)[0] for a in args[0]]
            return jnp.stack(res)

        return jax.pure_callback(qjit_func, shapes, args, vectorized=False)[0]

    return wrapper

def qjit(func):
    """JAX compatible qjit decorator"""

    @jax.custom_jvp
    def jax_func(*args):
        # Hardcoding in a qnode function with array(2,) output
        # and dtype float64. Questions:
        # - how do we determine this dynamically?
        # - how do we determine the batch dimensions?
        shapes = [jax.ShapeDtypeStruct((2,), jax.numpy.float64)]
        return _make_jax_jit(func, shapes)(*args)

    @jax_func.defjvp
    def f_jvp(primals, tangents):
        # hardcoding in a qnode function with tensor input of shape (3,)
        # and dtype float64. Questions:
        # - how do we determine this dynamically?
        # - how do we determine the batch dimensions?
        shapes = [jax.ShapeDtypeStruct((3, 2), jax.numpy.float64)]

        jac_fn = _make_jax_jit(catalyst.grad(func), shapes)
        jac = jac_fn(*primals)

        primals_out = jax_func(*primals)

        # Compute the vector jacobian products.
        # NOTE: this is assuming the QJIT function only has single array
        # output, and does not take into account returning multiple measurements
        tangents = jnp.squeeze(jnp.stack(tangents))

        if primals[0].ndim > 1:
            # batch dimension is present. Hardcoding in only *one* batch dim.
            tangents_out = jnp.stack([jnp.tensordot(j, t, axes=[0, 0]) for j, t in zip(jac, tangents)])
        else:
            tangents_out = jnp.tensordot(jac, tangents, axes=[[0], [0]])

        return primals_out, tangents_out

    return jax_func
```
</details>

We can see how seamlessly this allows `jax.jit`, `jax.grad`, and `qjit` to work together:

```python
import pennylane as qml
dev = qml.device("lightning.qubit", wires=1)

@qjit
@qml.qnode(dev)
def circuit(x):
    qml.RX(jnp.pi * x[0], wires=0)
    qml.RY(x[1] ** 2, wires=0)
    qml.RX(x[1] * x[2], wires=0)
    return qml.probs(wires=0)

@jax.jit
def cost_fn(weights):
    x = jnp.sin(weights)
    return jnp.sum(jnp.cos(circuit(x)) ** 2)
```

Executing the hybrid cost function, and computing the gradient:

```pycon
>>> weights = jnp.array([0.543, 0.123, 0.743])
>>> cost_fn(weights)
1.535381488575644
>>> jax.grad(cost_fn)(weights)
[-0.19387385 -0.04838516 -0.00651345]
```

The above also works for (_one_) batch dimension:

```python
>>> weights = jnp.array([[0.543, 0.123, 0.743], [0.1, 0.2, 0.3]])
>>> cost_fn(weights)
2.8580734398397474
>>> jax.grad(cost_fn)(weights)
[[-0.19387385 -0.04838516 -0.00651345]
 [ 0.49249037  0.05197949  0.02991883]]
```

**Additional questions**

- This examples shows why it would be important to add `catalyst.vjp` and `catalyst.jvp` transformations! It will be much more efficient to compute these quantities directly, rather than compute the full Jacobian and then post-process it. This may also be easier to do with Enzyme integration.

- I was surprised that vectorization doesn't seem to work, we should see if we can enable this. That is, 

  ```python
  @catalyst.qjit
  @qml.qnode(dev)
  def circuit(x):
      qml.RX(jnp.pi * x[0], wires=0)
      qml.RY(x[1] ** 2, wires=0)
      qml.RX(x[1] * x[2], wires=0)
      return qml.probs(wires=0)
  
  weights = jnp.array([[0.543, 0.123, 0.743], [0.1, 0.2, 0.3]])
  print(circuit(weights))
  ```

  ```
  File "/home/josh/xanadu/catalyst/frontend/catalyst/jax_primitives.py", line 471, in qinst_lowering
    if not ir.F64Type.isinstance(baseType):
  UnboundLocalError: local variable 'baseType' referenced before assignment
  ```

### PR DESCRIPTION
**Pull request by dime10**
Implements option (1) presented in issue #15 to enable JAX transformation support on qjitted functions.

[sc-34661]
[sc-37639]

Closes #15

### CODE REVIEW

File: frontend/catalyst/compilation_pipelines.py [hunk 1]
```diff
@@ -550,9 +546,87 @@ def __call__(self, *args, **kwargs):
...
+
+    def __init__(self, qfunc):
+        @functools.wraps(qfunc, assigned=("__annotations__",), updated=())
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
+
+            return catalyst.grad(qfunc, argnum=all_diff_args)(*args, **kwargs)
+
+        deriv_wrapper.__name__ = "deriv." + qfunc.__name__
```
Comment (erick-xanadu) [inline frontend/catalyst/compilation_pipelines.py:580]: Could this be expected to be the entry-point to the compiled function?

Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:580]: If I understand you correctly, yes. The `deriv_wrapper` is directly passed to `QJIT` in the next step to instantiate a new compiled function, making it the entry point of that newly compiled function.

Comment (erick-xanadu) [reply frontend/catalyst/compilation_pipelines.py:580]: I think we need to phase the `.` in `deriv.${name}` :sweat_smile: otherwise, linking the wrapper and catching the exception will involve a linker script.

Comment (erick-xanadu) [reply frontend/catalyst/compilation_pipelines.py:580]: To be more concrete, substitute the dot for an underscore.

Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:580]: Sounds good to me :)

Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py]: https://github.com/PennyLaneAI/catalyst/pull/96/commits/952d2eab74aa15b82be4f58f10b0f33ab0fc2b4e

File: frontend/catalyst/compilation_pipelines.py [hunk 2]
```diff
@@ -550,9 +546,102 @@ def __call__(self, *args, **kwargs):
...
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
+
+            return catalyst.grad(self.qfunc, argnum=all_diff_args)(*args, **kwargs)
+
+        annotations = {}
+        signature = inspect.signature(self.qfunc)
+        for idx, (arg_name, param) in enumerate(signature.parameters.items()):
+            param._annotation = self.qfunc.c_sig[idx]
```
Comment (dime10) [inline frontend/catalyst/compilation_pipelines.py:600]: Weird that pylint didn't warn about this when to code was located in the `__init__` function 🤔

Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py]: https://github.com/PennyLaneAI/catalyst/pull/96/commits/c1f649449c63472dc3deb449f7e5a7abcf386c38

File: frontend/catalyst/compilation_pipelines.py [hunk 3]
```diff
@@ -550,9 +546,102 @@ def __call__(self, *args, **kwargs):
...
+
+    @property
+    def deriv_qfunc(self):
+        """Compile a function computing the derivate of the wrapped QJIT function."""
+        if self._deriv_qfunc is not None:
+            return self._deriv_qfunc
+
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
```
Comment (maliasadi) [inline frontend/catalyst/compilation_pipelines.py:593]: Why non-floating args cannot be differentiable?  :thinking:

File: frontend/catalyst/compilation_pipelines.py (hunk 2 above)
Comment (maliasadi) [inline frontend/catalyst/compilation_pipelines.py]: What's exactly happening in this line? :)

File: frontend/catalyst/compilation_pipelines.py [hunk 4]
```diff
@@ -550,9 +546,107 @@ def __call__(self, *args, **kwargs):
...
+
+        self._deriv_qfunc = QJIT(deriv_wrapper, self.qfunc.compile_options)
+        return self._deriv_qfunc
+
+    @staticmethod
+    def wrap_callback(qfunc, *args, **kwargs):
+        """Wrap a QJIT function inside a jax host callback."""
+        return jax.pure_callback(qfunc, qfunc.jaxpr.out_avals, *args, vectorized=False, **kwargs)
+
+    def compute_jvp(self, primals, tangents):
```
Comment (maliasadi) [inline frontend/catalyst/compilation_pipelines.py:620]: nice implementation!

**Review by maliasadi — APPROVED** [review]
Thanks @dime10! Great addition! :100:

File: frontend/catalyst/compilation_pipelines.py (hunk 2 above)
Comment (maliasadi) [reply frontend/catalyst/compilation_pipelines.py]: find the answer! nwm :)

**Comment by maliasadi** [comment]
I may suggest updating the quick-start guideline including this new support in Catalyst.

**Comment by dime10** [comment]
> I may suggest updating the quick-start guideline including this new support in Catalyst.

Good idea!

File: frontend/test/pytest/test_jax_integration.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            qml.RX(x[1] * x[2], wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        @jax.jit
+        def cost_fn(x):
+            result = circuit(x)
+            return jnp.cos(result) ** 2
+
+        result = cost_fn(jnp.array([0.1, 0.2, 0.3]))
+        assert jnp.allclose(result, 0.3573519347253657)
```
Comment (josh146) [inline frontend/test/pytest/test_jax_integration.py:47]: @dime10 I recommend using analytic results for the comparison. E.g., for this circuit, and the circuits below, the analytic derivative should be pretty simple (a basic trig function).

We make a deliberate effort to do this in PL, because there were quite a few times bugs were not caught because 'magic numbers' used in test comparisons were computed incorrectly, whereas analytic results are easy to check

Comment (josh146) [reply frontend/test/pytest/test_jax_integration.py:47]: You can pilfer the [pennylane jax tests](https://github.com/PennyLaneAI/pennylane/blob/master/tests/interfaces/test_jax_qnode.py) for some examples :)

File: frontend/test/pytest/test_jax_integration.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+    def test_multiple_arguments(self):
+        """Test a circuit with multiple arguments using jax.jit on top of qjit."""
```
Comment (josh146) [inline frontend/test/pytest/test_jax_integration.py:50]: nice!

File: frontend/test/pytest/test_jax_integration.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+        @qjit
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def circuit(x: jax.ShapedArray((3,), dtype=float), y: jax.ShapedArray((2,), dtype=float)):
+            qml.RX(jnp.pi * x[0], wires=0)
+            qml.RY(x[1] ** 2, wires=0)
+            qml.RX(y[1] * x[2], wires=0)
+            return qml.probs(wires=0)
```
Comment (josh146) [inline frontend/test/pytest/test_jax_integration.py:58]: this is a good example as well because the circuit is vector valued.

File: frontend/test/pytest/test_jax_integration.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        def cost_fn(x, y):
+            m1 = circuit(x)
+            m2 = circuit(y)
+            return m1 + m2
+
+        result1 = cost_fn(0.0, 0.0)
+        result2 = cost_fn(0.0, jnp.pi)
+        assert jnp.allclose(result1, 0.0)
+        assert jnp.allclose(result2, 0.0)
+
```
Comment (josh146) [inline frontend/test/pytest/test_jax_integration.py:250]: @dime10 it might be worth also adding tests for:

- `jax.jacobian` with a vector valued QNode
- `jax.vmap`, assuming it is working?

File: frontend/test/pytest/test_tracing.py [hunk 9]
```diff
@@ -92,16 +91,6 @@ def repeat_until_false(i, _):
...
 
-    def test_tracing_through_qjit(self):
-        """Check for useful error message when JAX is used to trace through a qjit function."""
-
-        @qjit
-        def func(x):
-            return x**2
-
-        with pytest.raises(ValueError, match="Cannot use JAX to trace through a qjit"):
-            jax.grad(func)(2.0)
```
Comment (josh146) [inline frontend/test/pytest/test_tracing.py:103]: 🎉

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks @dime10, this is super cool 😍 Just a few comments and suggestions

File: frontend/catalyst/compilation_pipelines.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:593]: Well typically you don't define derivatives on integer or Boolean-valued functions because the function must be continuous and smooth. You can do complex differentiation but I don't think it's supported in Catalyst right now.

The reason I compute the indices in this way here is to be able to support functions that have differentiable as well as non-differentiable arguments (e.g. `def f(x: float, y: int)`). If I didn't exclude the non-differentiable ones, Catalyst would raise an error at this stage.

File: frontend/catalyst/compilation_pipelines.py [hunk 10]
```diff
@@ -550,9 +546,107 @@ def __call__(self, *args, **kwargs):
...
+            return results
+
+        self.qfunc = qfunc
+        self.jaxed_qfunc = jaxed_qfunc
+        self._deriv_qfunc = None  # compiled on demand
+        jaxed_qfunc.defjvp(self.compute_jvp)
+
+    @property
+    def deriv_qfunc(self):
+        """Compile a function computing the derivate of the wrapped QJIT function."""
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:587]: This always happens to me I don't know why 😆

File: frontend/catalyst/compilation_pipelines.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:593]: @josh146 Yes, any tensor with dtype starting with "f".

File: doc/changelog.md [hunk 11]
```diff
@@ -2,6 +2,13 @@
 
 <h3>New features</h3>
 
+* Support for JAX transformations on top of ``catalyst.qjit``.
+  [#96](https://github.com/PennyLaneAI/catalyst/pull/96)
+
+  Now supported:
+   * jitting with JAX
+   * first-order differentiation in both forward and reverse mode
```
[the remark this answers is not in the document]
Comment (dime10) [reply doc/changelog.md:10]: I think was thinking of adding vmap in a separate PR, but might as well do it here 😅

File: frontend/catalyst/compilation_pipelines.py (hunk 3 above)
Comment (maliasadi) [reply frontend/catalyst/compilation_pipelines.py:593]: Could users still specify the differentiable arguments? Isn't the assumption of considering all floating points arguments as differentiable args impacting the user experience? Wouldn't it be good to print a warning for cases with non-floating args?

File: doc/changelog.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (josh146) [reply doc/changelog.md:10]: oh that's fine! probably cleaner/better to do it in a new PR

File: frontend/catalyst/compilation_pipelines.py [hunk 12]
```diff
@@ -550,9 +546,107 @@ def __call__(self, *args, **kwargs):
...
+        if self._deriv_qfunc is not None:
+            return self._deriv_qfunc
+
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
+
+            return catalyst.grad(self.qfunc, argnum=all_diff_args)(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:595]: That should still work, although inefficiently. Catalyst will compute the Jacobian for all arguments and sum up JVPs across arguments to produce the `custom_jvp` result. 

Originally, I didn't know if there was a way to determine how the `argnums` (which user may have specified at a higher level) propagated down to the arguments of our function, so this was my first attempt to get it working.
I'm not 100% sure, but if I'm not mistaken arguments that don't participate in differentiation will result in simply having a tangent vector of all zeros for that argument. My assertion here encodes this assumption:
https://github.com/PennyLaneAI/catalyst/blob/c1f649449c63472dc3deb449f7e5a7abcf386c38/frontend/catalyst/compilation_pipelines.py#L628-L629
In any case, a tangent vector of all zeros indicates that the corresponding argument can be skipped for Jacobian computation since the Jacobian-vector product will be zero and thus contribute nothing to the final sum.
 
I actually quickly prototyped this yesterday as follows:
```py
        # Optimization: Do not compute Jacobians for tangents which do not participate in
        #               differentiation.
        # TODO: replace with symbolic zero support
        argnums = []
        for idx, tangent in enumerate(tangents):
            if tangent.dtype.kind == "f":
                if is_abstract(tangent, like="jax"):
                    argnums.append(idx)
                elif any(jnp.atleast_1d(tangent != 0.0)):
                    argnums.append(idx)
        derivatives = self.wrap_callback(self.get_derivative_qfunc(argnums), *primals)
```
and
```py
    def get_derivative_qfunc(self, argnums):
        """Get or generate a derivative version of the wrapped QJIT for a given set of argnums."""

        argnum_key = "".join(str(idx) for idx in argnums)
        if not argnum_key in self.deriv_qfuncs:

            @functools.wraps(self.qfunc, assigned=("__annotations__",), updated=())
            def deriv_wrapper(*args, **kwargs):
                # Only consider differentiable arguments (i.e. of type floating point).
                if any(args[idx].dtype.kind != "f" for idx in argnums):
                    raise ValueError("Attempted to differentiate non-floating point argument!")

                return catalyst.grad(self.qfunc, argnum=argnums)(*args, **kwargs)

            deriv_wrapper.__name__ = "deriv_" + argnum_key + "." + self.qfunc.__name__
            self.deriv_qfuncs[argnum_key] = QJIT(deriv_wrapper, self.qfunc.compile_options)

        return self.deriv_qfuncs[argnum_key]
```
One downside is that we compile more often (every time a new unique set of argnums is encountered), but that should be well worth it for the savings in gradient compute time.

The above can be improved further with something @rmoyard pointed out, which is that JAX recently implemented [symbolic zeros](https://github.com/google/jax/issues/11208) for tangent vectors. Once we upgrade our JAX version we could make use of it here.

File: frontend/catalyst/compilation_pipelines.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:593]: > Could users still specify the differentiable arguments? Isn't the assumption of considering all floating points arguments as differentiable args impacting the user experience?

The user specifies argnums at a different level to which we don't have direct access, see my discussion of the issue [under Josh's comment](https://github.com/PennyLaneAI/catalyst/pull/96#discussion_r1173162649). It's mainly a question of performance however, the user experience should not be affected.

> Wouldn't it be good to print a warning for cases with non-floating args?

In the case that a user does somehow define a qjit function on complex numbers, and they also attempt to differentiate it using JAX with holomorphic gradient enabled (which I think this is pretty rare), it would be good to raise an error. I think this means checking whether a tangent vector for which we didn't compute a Jacobian contains non-zero values, which I actually do here:
https://github.com/PennyLaneAI/catalyst/blob/c1f649449c63472dc3deb449f7e5a7abcf386c38/frontend/catalyst/compilation_pipelines.py#L628-L629

File: frontend/test/pytest/test_jax_integration.py (hunk 5 above)
Comment (dime10) [reply frontend/test/pytest/test_jax_integration.py:47]: > We make a deliberate effort to do this in PL, because there were quite a few times bugs were not caught because 'magic numbers' used in test comparisons were computed incorrectly, whereas analytic results are easy to check

That's a fair point. My thinking was to follow your advice of not comparing against executions of other frameworks (e.g. comparing to the result produced by PennyLane every time), but a magic number might actually be worse 😅

File: frontend/test/pytest/test_jax_integration.py (hunk 8 above)
Comment (dime10) [reply frontend/test/pytest/test_jax_integration.py:250]: Good idea, will add!

Comment (dime10) [reply frontend/test/pytest/test_jax_integration.py:279]: https://github.com/PennyLaneAI/catalyst/pull/96/commits/881669a5662d3f9516e088fc921fb9311298449f

File: frontend/catalyst/compilation_pipelines.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py]: I implemented the optimization here: https://github.com/PennyLaneAI/catalyst/pull/96/commits/f89faf6b71d41069e2f44593b4edf9fd03a36916

File: frontend/catalyst/compilation_pipelines.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py]: https://github.com/PennyLaneAI/catalyst/pull/96/commits/8fd0c6d289119d7bc6d52b840a8990cfb0bc3154

File: frontend/catalyst/compilation_pipelines.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py]: After implementing the optimization in https://github.com/PennyLaneAI/catalyst/pull/96/commits/f89faf6b71d41069e2f44593b4edf9fd03a36916 this is no longer an issue, we can let the grad op itself raise an error for incorrect types (added here: https://github.com/PennyLaneAI/catalyst/pull/96/commits/efb78c001ae619538d75d86f3590ed3eef3496da, previously those would be raised in MLIR instead).

File: doc/changelog.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (dime10) [reply doc/changelog.md]: > @dime10 do you mind adding a small example here to demonstrate to an interested user what the new feature is? We do this with the PL changelog, and as development speed increases over time, we find it really helps keep track of what new features mean in practice for users

Changed here: https://github.com/PennyLaneAI/catalyst/pull/96/commits/25a3ad3a6b59e8036a98f16aa619e45ab611229e

File: Makefile [hunk 13]
```diff
@@ -90,7 +90,7 @@ clean-all:
 .PHONY: coverage coverage-frontend coverage-runtime
 coverage: coverage-frontend coverage-runtime
 
-coverage-frontend: | frontend
+coverage-frontend:
```
Comment (erick-xanadu) [inline Makefile:93]: Ah, I see you also encountered this.

**Review by erick-xanadu — APPROVED** [review]

File: frontend/test/pytest/test_jax_integration.py (hunk 5 above)
Comment (josh146) [reply frontend/test/pytest/test_jax_integration.py:47]: Yeah, we've had issues in the past with both comparisons to other frameworks _and_ to magic numbers 😅 Analytic comparisons though have worked really well.

Comment (dime10) [reply frontend/test/pytest/test_jax_integration.py]: I removed all comparisons to magic numbers here: https://github.com/PennyLaneAI/catalyst/pull/96/commits/cd18477afd34914a5a70e93df00285078554b8cc

File: doc/changelog.md [hunk 14]
```diff
@@ -2,6 +2,53 @@
 
 <h3>New features</h3>
 
+* Catalyst programs can now be used inside of a larger JAX workflow which uses JIT compilation,
+  automatic differentiation, and other JAX transforms.
+  [#96](https://github.com/PennyLaneAI/catalyst/pull/96)
+
+  Note that generally Catalyst should be used to JIT the entire workflow, but sometimes users may
+  wish to delegate only the quantum part of their workflow to Catalyst and let JAX handle the rest
+  (for example due to missing a feature or compatibility issue in Catalyst).
```
Comment (maliasadi) [inline doc/changelog.md:11]: Would you like to add this new support as a note to quick-start guideline?

File: frontend/test/pytest/test_gradient.py [hunk 15]
```diff
@@ -688,5 +690,42 @@ def interpret_grad(params):
...
+    """Test a function which attempts to differentiate non-floating point results."""
+
+    @qml.qnode(qml.device("lightning.qubit", wires=2))
+    def circuit(x: float, y: float):
+        qml.RX(x, wires=0)
+        qml.RY(y, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    @qjit
+    @grad
```
Comment (maliasadi) [inline frontend/test/pytest/test_gradient.py:722]: Weird use of the grad op as a decorator :)

File: frontend/test/pytest/test_jax_integration.py [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+
+class TestJAXJIT:
+    """Test QJIT compatibility with JAX compilation."""
+
+    def test_simple_circuit(self):
+        """Test a basic use case of jax.jit on top of qjit."""
+
+        @qjit
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
```
Comment (maliasadi) [inline frontend/test/pytest/test_jax_integration.py:34]: Could you also update these tests after merging with the main branch as follows?
```suggestion
    def test_simple_circuit(self, backend):
        """Test a basic use case of jax.jit on top of qjit."""

        @qjit
        @qml.qnode(qml.device(backend, wires=2))
```

File: frontend/catalyst/compilation_pipelines.py [hunk 17]
```diff
@@ -550,9 +546,111 @@ def __call__(self, *args, **kwargs):
...
+
+        self.deriv_qfuncs[argnum_key] = QJIT(deriv_wrapper, self.qfunc.compile_options)
+        return self.deriv_qfuncs[argnum_key]
+
+    def compute_jvp(self, primals, tangents):
+        """Compute the set of results and JVPs for a QJIT function."""
+
+        # Optimization: Do not compute Jacobians for arguments which do not participate in
+        #               differentiation.
+        # TODO: replace with symbolic zero support
```
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/compilation_pipelines.py:637]: [perhaps open github issues?]

File: frontend/catalyst/pennylane_extensions.py [hunk 18]
```diff
@@ -181,6 +183,19 @@ def __call__(self, *args, **kwargs):
...
+                raise TypeError(
+                    "Catalyst.grad only supports differentiation on floating-point "
+                    f"arguments, got '{arg.dtype}' at position {pos}."
+                )
+        for pos, res in enumerate(jaxpr.out_avals):
+            if res.dtype.kind != "f":
+                raise TypeError(
+                    "Catalyst.grad only supports differentiation on floating-point "
+                    f"results, got '{res.dtype}' at position {pos}."
+                )
```
Comment (josh146) [inline frontend/catalyst/pennylane_extensions.py:197]: Did the JAX integration unearth some edge cases/bugs?

File: frontend/test/pytest/test_gradient.py (hunk 15 above)
Comment (josh146) [reply frontend/test/pytest/test_gradient.py:722]: 😆

Comment (josh146) [reply frontend/test/pytest/test_gradient.py:722]: I sometimes do this, if you don't care about the original cost function and only need the grad

**Review by josh146 — APPROVED** [review]
Nice work @dime10 :)

File: frontend/catalyst/compilation_pipelines.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:637]: Yes I think GitHub issues would be best for these small TODOs

[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:637]: https://github.com/PennyLaneAI/catalyst/issues/125

File: frontend/catalyst/compilation_pipelines.py [hunk 19]
```diff
@@ -550,9 +546,111 @@ def __call__(self, *args, **kwargs):
...
+            for res_idx in range(len(results)):
+                deriv_idx = arg_idx * len(results) + res_idx
+                num_axes = 0 if tangent.ndim == 0 else 1
+                jvp = jnp.tensordot(jnp.transpose(derivatives[deriv_idx]), tangent, axes=num_axes)
+                jvps[res_idx] = jvps[res_idx] + jvp
+
+        if len(results) == 1:
+            results = results[0]
+        if len(jvps) == 1:
+            jvps = jvps[0]
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:660]: Yeah I think if you consistently used JAX's PyTree infrastructure with the flattening and unflattening functions this wouldn't be an issue, so this could definitely be improved.

File: frontend/test/pytest/test_jax_integration.py (hunk 16 above)
Comment (dime10) [reply frontend/test/pytest/test_jax_integration.py:34]: Thanks, will do!

Comment (dime10) [reply frontend/test/pytest/test_jax_integration.py:34]: a521c48c2d27054bc58226fa5937f2bc8791ec08

File: doc/changelog.md (hunk 14 above)
Comment (dime10) [reply doc/changelog.md:11]: https://github.com/PennyLaneAI/catalyst/pull/96/commits/e191f8332b318146bea666912fe98b50712c8c22
